File: Core/Models/PhysicalMetric.swift
import Foundation

public enum BodySide: String, Codable, CaseIterable, Sendable, Hashable {
    case left, right
    public var labelKey: String { "body_side.\(rawValue)" }
}

public enum PhysicalCategory: String, Codable, CaseIterable, Sendable, Hashable {
    case flexibility, power, speed, endurance, strength, bodyComposition

    public var labelKey: String { "physical_category.\(rawValue)" }

    public var systemIcon: String {
        switch self {
        case .flexibility: "figure.flexibility"
        case .power: "bolt.fill"
        case .speed: "hare.fill"
        case .endurance: "lungs.fill"
        case .strength: "dumbbell.fill"
        case .bodyComposition: "scalemass.fill"
        }
    }
}

public enum TestFrequency: String, Codable, CaseIterable, Sendable, Hashable {
    case weekly, monthly, quarterly

    public var labelKey: String { "test_frequency.\(rawValue)" }

    /// Days between recommended captures.
    public var dayInterval: Int {
        switch self {
        case .weekly: 7
        case .monthly: 30
        case .quarterly: 90
        }
    }

    /// Past this many days beyond `dayInterval`, the test is considered overdue
    /// (red rather than amber).
    public var graceDays: Int {
        switch self {
        case .weekly: 3
        case .monthly: 7
        case .quarterly: 14
        }
    }
}

public enum PhysicalMetricKind: String, Codable, CaseIterable, Sendable, Hashable {
    // Flexibility
    case frontSplitCm
    case sideSplitCm
    case kickHeightHead
    case kickHeightChest
    case legRaiseAngle

    // Power
    case roundhouseForceBack
    case roundhouseForceFront
    case sideKickForce
    case verticalJumpCm
    case broadJumpCm

    // Speed
    case roundhouseKicks10s
    case backKicks10s
    case sprint20mSec
    case reactionMs

    // Endurance
    case yoyoLevel
    case twoMinKickTotal
    case plankSec

    // Strength
    case singleLegSquatReps
    case hollowBodyHoldSec
    case pushUps60s

    // Body composition
    case bodyWeightKg
    case bodyFatPct
    case restingHr

    public var labelKey: String { "metric.\(rawValue)" }

    public var category: PhysicalCategory {
        switch self {
        case .frontSplitCm, .sideSplitCm, .kickHeightHead, .kickHeightChest, .legRaiseAngle:
            .flexibility
        case .roundhouseForceBack, .roundhouseForceFront, .sideKickForce, .verticalJumpCm, .broadJumpCm:
            .power
        case .roundhouseKicks10s, .backKicks10s, .sprint20mSec, .reactionMs:
            .speed
        case .yoyoLevel, .twoMinKickTotal, .plankSec:
            .endurance
        case .singleLegSquatReps, .hollowBodyHoldSec, .pushUps60s:
            .strength
        case .bodyWeightKg, .bodyFatPct, .restingHr:
            .bodyComposition
        }
    }

    public var frequency: TestFrequency {
        switch self {
        case .frontSplitCm, .sideSplitCm, .kickHeightHead, .kickHeightChest: .monthly
        case .legRaiseAngle: .quarterly
        case .roundhouseForceBack, .roundhouseForceFront, .sideKickForce: .quarterly
        case .verticalJumpCm, .broadJumpCm: .quarterly
        case .roundhouseKicks10s, .backKicks10s: .monthly
        case .sprint20mSec, .reactionMs: .quarterly
        case .yoyoLevel: .quarterly
        case .twoMinKickTotal, .plankSec: .monthly
        case .singleLegSquatReps, .hollowBodyHoldSec, .pushUps60s: .quarterly
        case .bodyWeightKg, .restingHr: .weekly
        case .bodyFatPct: .monthly
        }
    }

    /// Display unit; empty string for pass/fail.
    public var unit: String {
        switch self {
        case .frontSplitCm, .sideSplitCm, .verticalJumpCm, .broadJumpCm: "cm"
        case .kickHeightHead, .kickHeightChest: ""
        case .legRaiseAngle: "°"
        case .roundhouseForceBack, .roundhouseForceFront, .sideKickForce: "kg"
        case .roundhouseKicks10s, .backKicks10s, .twoMinKickTotal: "kicks"
        case .sprint20mSec, .plankSec, .hollowBodyHoldSec: "s"
        case .reactionMs: "ms"
        case .yoyoLevel: "lvl"
        case .singleLegSquatReps, .pushUps60s: "reps"
        case .bodyWeightKg: "kg"
        case .bodyFatPct: "%"
        case .restingHr: "bpm"
        }
    }

    public var isPassFail: Bool {
        switch self {
        case .kickHeightHead, .kickHeightChest: true
        default: false
        }
    }

    /// Recorded once per leg (left + right are separate rows).
    public var isUnilateral: Bool {
        switch self {
        case .singleLegSquatReps: true
        default: false
        }
    }

    /// Stepper / slider hints (lower, upper, step).
    public var inputRange: (lower: Double, upper: Double, step: Double) {
        switch self {
        case .frontSplitCm, .sideSplitCm: (0, 60, 0.5)
        case .kickHeightHead, .kickHeightChest: (0, 1, 1)
        case .legRaiseAngle: (0, 180, 1)
        case .roundhouseForceBack, .roundhouseForceFront, .sideKickForce: (0, 600, 1)
        case .verticalJumpCm: (0, 80, 1)
        case .broadJumpCm: (0, 350, 1)
        case .roundhouseKicks10s, .backKicks10s: (0, 60, 1)
        case .sprint20mSec: (2.0, 6.0, 0.01)
        case .reactionMs: (100, 500, 1)
        case .yoyoLevel: (1, 21, 1)
        case .twoMinKickTotal: (0, 300, 1)
        case .plankSec, .hollowBodyHoldSec: (0, 300, 1)
        case .singleLegSquatReps, .pushUps60s: (0, 100, 1)
        case .bodyWeightKg: (15, 140, 0.1)
        case .bodyFatPct: (3, 50, 0.1)
        case .restingHr: (30, 120, 1)
        }
    }

    /// True if a higher raw number means a better result. Used for normalising
    /// the composite score and for trend-direction icons.
    public var higherIsBetter: Bool {
        switch self {
        case .sprint20mSec, .reactionMs,
             .frontSplitCm, .sideSplitCm,
             .bodyFatPct, .restingHr: false
        default: true
        }
    }

    /// 0...1 normalisation against the input range. Used by the composite.
    public func normalized(_ value: Double) -> Double {
        let r = inputRange
        guard r.upper > r.lower else { return 0 }
        let raw = (value - r.lower) / (r.upper - r.lower)
        let clamped = max(0, min(1, raw))
        return higherIsBetter ? clamped : 1 - clamped
    }
}

public struct PhysicalMetric: Codable, Identifiable, Hashable, Sendable {
    public let id: EntityID
    public var athleteID: EntityID
    public var recordedAt: Date
    public var recordedByCoachID: EntityID
    public var kind: PhysicalMetricKind
    /// Raw measurement in the kind's unit. For pass/fail kinds: 0 = fail, 1 = pass.
    public var value: Double
    /// Set only for unilateral kinds (e.g. single-leg squat). nil otherwise.
    public var leg: BodySide?
    public var notes: String?

    public init(
        id: EntityID = UUID(),
        athleteID: EntityID,
        recordedAt: Date,
        recordedByCoachID: EntityID,
        kind: PhysicalMetricKind,
        value: Double,
        leg: BodySide? = nil,
        notes: String? = nil
    ) {
        self.id = id
        self.athleteID = athleteID
        self.recordedAt = recordedAt
        self.recordedByCoachID = recordedByCoachID
        self.kind = kind
        self.value = value
        self.leg = leg
        self.notes = notes
    }
}

public extension Array where Element == PhysicalMetric {
    /// Latest entry per (kind, leg) tuple. Drives the dashboard "current values".
    func latestPerKind() -> [PhysicalMetric] {
        var seen: [String: PhysicalMetric] = [:]
        for m in self.sorted(by: { $0.recordedAt > $1.recordedAt }) {
            let key = "\(m.kind.rawValue)|\(m.leg?.rawValue ?? "-")"
            if seen[key] == nil { seen[key] = m }
        }
        return Array(seen.values)
    }
}

File: flutter_app/lib/core/models/physical_metric.dart
import 'entity_id.dart';

/// Port of Core/Models/PhysicalMetric.swift.

enum BodySide {
  left,
  right;

  String get labelKey => 'body_side.$name';

  static BodySide fromJson(String raw) => BodySide.values.firstWhere(
    (e) => e.name == raw,
    orElse: () => BodySide.left,
  );
}

enum PhysicalCategory {
  flexibility,
  power,
  speed,
  endurance,
  strength,
  bodyComposition;

  String get labelKey => 'physical_category.$name';

  String get systemIcon => switch (this) {
    PhysicalCategory.flexibility => 'figure.flexibility',
    PhysicalCategory.power => 'bolt.fill',
    PhysicalCategory.speed => 'hare.fill',
    PhysicalCategory.endurance => 'lungs.fill',
    PhysicalCategory.strength => 'dumbbell.fill',
    PhysicalCategory.bodyComposition => 'scalemass.fill',
  };
}

enum TestFrequency {
  weekly,
  monthly,
  quarterly;

  String get labelKey => 'test_frequency.$name';

  /// Days between recommended captures.
  int get dayInterval => switch (this) {
    TestFrequency.weekly => 7,
    TestFrequency.monthly => 30,
    TestFrequency.quarterly => 90,
  };

  /// Past this many days beyond [dayInterval], the test is considered overdue.
  int get graceDays => switch (this) {
    TestFrequency.weekly => 3,
    TestFrequency.monthly => 7,
    TestFrequency.quarterly => 14,
  };
}

enum PhysicalMetricKind {
  // Flexibility
  frontSplitCm,
  sideSplitCm,
  kickHeightHead,
  kickHeightChest,
  legRaiseAngle,

  // Power
  roundhouseForceBack,
  roundhouseForceFront,
  sideKickForce,
  verticalJumpCm,
  broadJumpCm,

  // Speed
  roundhouseKicks10s,
  backKicks10s,
  sprint20mSec,
  reactionMs,

  // Endurance
  yoyoLevel,
  twoMinKickTotal,
  plankSec,

  // Strength
  singleLegSquatReps,
  hollowBodyHoldSec,
  pushUps60s,

  // Body composition
  bodyWeightKg,
  bodyFatPct,
  restingHr;

  String get labelKey => 'metric.$name';

  PhysicalCategory get category => switch (this) {
    PhysicalMetricKind.frontSplitCm ||
    PhysicalMetricKind.sideSplitCm ||
    PhysicalMetricKind.kickHeightHead ||
    PhysicalMetricKind.kickHeightChest ||
    PhysicalMetricKind.legRaiseAngle => PhysicalCategory.flexibility,
    PhysicalMetricKind.roundhouseForceBack ||
    PhysicalMetricKind.roundhouseForceFront ||
    PhysicalMetricKind.sideKickForce ||
    PhysicalMetricKind.verticalJumpCm ||
    PhysicalMetricKind.broadJumpCm => PhysicalCategory.power,
    PhysicalMetricKind.roundhouseKicks10s ||
    PhysicalMetricKind.backKicks10s ||
    PhysicalMetricKind.sprint20mSec ||
    PhysicalMetricKind.reactionMs => PhysicalCategory.speed,
    PhysicalMetricKind.yoyoLevel ||
    PhysicalMetricKind.twoMinKickTotal ||
    PhysicalMetricKind.plankSec => PhysicalCategory.endurance,
    PhysicalMetricKind.singleLegSquatReps ||
    PhysicalMetricKind.hollowBodyHoldSec ||
    PhysicalMetricKind.pushUps60s => PhysicalCategory.strength,
    PhysicalMetricKind.bodyWeightKg ||
    PhysicalMetricKind.bodyFatPct ||
    PhysicalMetricKind.restingHr => PhysicalCategory.bodyComposition,
  };

  TestFrequency get frequency => switch (this) {
    PhysicalMetricKind.frontSplitCm ||
    PhysicalMetricKind.sideSplitCm ||
    PhysicalMetricKind.kickHeightHead ||
    PhysicalMetricKind.kickHeightChest => TestFrequency.monthly,
    PhysicalMetricKind.legRaiseAngle => TestFrequency.quarterly,
    PhysicalMetricKind.roundhouseForceBack ||
    PhysicalMetricKind.roundhouseForceFront ||
    PhysicalMetricKind.sideKickForce => TestFrequency.quarterly,
    PhysicalMetricKind.verticalJumpCm ||
    PhysicalMetricKind.broadJumpCm => TestFrequency.quarterly,
    PhysicalMetricKind.roundhouseKicks10s ||
    PhysicalMetricKind.backKicks10s => TestFrequency.monthly,
    PhysicalMetricKind.sprint20mSec ||
    PhysicalMetricKind.reactionMs => TestFrequency.quarterly,
    PhysicalMetricKind.yoyoLevel => TestFrequency.quarterly,
    PhysicalMetricKind.twoMinKickTotal ||
    PhysicalMetricKind.plankSec => TestFrequency.monthly,
    PhysicalMetricKind.singleLegSquatReps ||
    PhysicalMetricKind.hollowBodyHoldSec ||
    PhysicalMetricKind.pushUps60s => TestFrequency.quarterly,
    PhysicalMetricKind.bodyWeightKg ||
    PhysicalMetricKind.restingHr => TestFrequency.weekly,
    PhysicalMetricKind.bodyFatPct => TestFrequency.monthly,
  };

  /// Display unit; empty string for pass/fail.
  String get unit => switch (this) {
    PhysicalMetricKind.frontSplitCm ||
    PhysicalMetricKind.sideSplitCm ||
    PhysicalMetricKind.verticalJumpCm ||
    PhysicalMetricKind.broadJumpCm => 'cm',
    PhysicalMetricKind.kickHeightHead ||
    PhysicalMetricKind.kickHeightChest => '',
    PhysicalMetricKind.legRaiseAngle => '°',
    PhysicalMetricKind.roundhouseForceBack ||
    PhysicalMetricKind.roundhouseForceFront ||
    PhysicalMetricKind.sideKickForce => 'kg',
    PhysicalMetricKind.roundhouseKicks10s ||
    PhysicalMetricKind.backKicks10s ||
    PhysicalMetricKind.twoMinKickTotal => 'kicks',
    PhysicalMetricKind.sprint20mSec ||
    PhysicalMetricKind.plankSec ||
    PhysicalMetricKind.hollowBodyHoldSec => 's',
    PhysicalMetricKind.reactionMs => 'ms',
    PhysicalMetricKind.yoyoLevel => 'lvl',
    PhysicalMetricKind.singleLegSquatReps ||
    PhysicalMetricKind.pushUps60s => 'reps',
    PhysicalMetricKind.bodyWeightKg => 'kg',
    PhysicalMetricKind.bodyFatPct => '%',
    PhysicalMetricKind.restingHr => 'bpm',
  };

  bool get isPassFail => switch (this) {
    PhysicalMetricKind.kickHeightHead ||
    PhysicalMetricKind.kickHeightChest => true,
    _ => false,
  };

  /// Recorded once per leg (left + right are separate rows).
  bool get isUnilateral => switch (this) {
    PhysicalMetricKind.singleLegSquatReps => true,
    _ => false,
  };

  /// Stepper/slider hints as (lower, upper, step).
  (double lower, double upper, double step) get inputRange => switch (this) {
    PhysicalMetricKind.frontSplitCm ||
    PhysicalMetricKind.sideSplitCm => (0, 60, 0.5),
    PhysicalMetricKind.kickHeightHead ||
    PhysicalMetricKind.kickHeightChest => (0, 1, 1),
    PhysicalMetricKind.legRaiseAngle => (0, 180, 1),
    PhysicalMetricKind.roundhouseForceBack ||
    PhysicalMetricKind.roundhouseForceFront ||
    PhysicalMetricKind.sideKickForce => (0, 600, 1),
    PhysicalMetricKind.verticalJumpCm => (0, 80, 1),
    PhysicalMetricKind.broadJumpCm => (0, 350, 1),
    PhysicalMetricKind.roundhouseKicks10s ||
    PhysicalMetricKind.backKicks10s => (0, 60, 1),
    PhysicalMetricKind.sprint20mSec => (2.0, 6.0, 0.01),
    PhysicalMetricKind.reactionMs => (100, 500, 1),
    PhysicalMetricKind.yoyoLevel => (1, 21, 1),
    PhysicalMetricKind.twoMinKickTotal => (0, 300, 1),
    PhysicalMetricKind.plankSec ||
    PhysicalMetricKind.hollowBodyHoldSec => (0, 300, 1),
    PhysicalMetricKind.singleLegSquatReps ||
    PhysicalMetricKind.pushUps60s => (0, 100, 1),
    PhysicalMetricKind.bodyWeightKg => (15, 140, 0.1),
    PhysicalMetricKind.bodyFatPct => (3, 50, 0.1),
    PhysicalMetricKind.restingHr => (30, 120, 1),
  };

  /// True if a higher raw number means a better result.
  bool get higherIsBetter => switch (this) {
    PhysicalMetricKind.sprint20mSec ||
    PhysicalMetricKind.reactionMs ||
    PhysicalMetricKind.frontSplitCm ||
    PhysicalMetricKind.sideSplitCm ||
    PhysicalMetricKind.bodyFatPct ||
    PhysicalMetricKind.restingHr => false,
    _ => true,
  };

  /// 0…1 normalisation against the input range.
  double normalized(double value) {
    final r = inputRange; // positional record: (lower $1, upper $2, step $3)
    if (r.$2 <= r.$1) return 0;
    final raw = (value - r.$1) / (r.$2 - r.$1);
    final clamped = raw.clamp(0.0, 1.0);
    return higherIsBetter ? clamped : 1 - clamped;
  }

  static PhysicalMetricKind fromJson(String raw) =>
      PhysicalMetricKind.values.firstWhere(
        (e) => e.name == raw,
        orElse: () => PhysicalMetricKind.bodyWeightKg,
      );
}

class PhysicalMetric {
  final EntityID id;
  final EntityID athleteId;
  final DateTime recordedAt;
  final EntityID recordedByCoachId;
  final PhysicalMetricKind kind;

  /// Raw measurement in the kind's unit. For pass/fail kinds: 0 = fail, 1 = pass.
  final double value;

  /// Set only for unilateral kinds (e.g. single-leg squat). null otherwise.
  final BodySide? leg;
  final String? notes;

  const PhysicalMetric({
    required this.id,
    required this.athleteId,
    required this.recordedAt,
    required this.recordedByCoachId,
    required this.kind,
    required this.value,
    this.leg,
    this.notes,
  });

  factory PhysicalMetric.create({
    EntityID? id,
    required EntityID athleteId,
    required DateTime recordedAt,
    required EntityID recordedByCoachId,
    required PhysicalMetricKind kind,
    required double value,
    BodySide? leg,
    String? notes,
  }) => PhysicalMetric(
    id: id ?? newEntityId(),
    athleteId: athleteId,
    recordedAt: recordedAt,
    recordedByCoachId: recordedByCoachId,
    kind: kind,
    value: value,
    leg: leg,
    notes: notes,
  );

  factory PhysicalMetric.fromJson(Map<String, dynamic> json) => PhysicalMetric(
    id: json['id'] as String,
    athleteId: json['athleteID'] as String,
    recordedAt: DateTime.parse(json['recordedAt'] as String),
    recordedByCoachId: json['recordedByCoachID'] as String,
    kind: PhysicalMetricKind.fromJson(json['kind'] as String),
    value: (json['value'] as num).toDouble(),
    leg: json['leg'] != null ? BodySide.fromJson(json['leg'] as String) : null,
    notes: json['notes'] as String?,
  );

  Map<String, dynamic> toJson() => {
    'id': id,
    'athleteID': athleteId,
    'recordedAt': recordedAt.toIso8601String(),
    'recordedByCoachID': recordedByCoachId,
    'kind': kind.name,
    'value': value,
    'leg': leg?.name,
    'notes': notes,
  };
}

extension PhysicalMetricListExt on List<PhysicalMetric> {
  /// Latest entry per (kind, leg) tuple. Drives the dashboard "current values".
  List<PhysicalMetric> latestPerKind() {
    final seen = <String, PhysicalMetric>{};
    final sorted = [...this]
      ..sort((a, b) => b.recordedAt.compareTo(a.recordedAt));
    for (final m in sorted) {
      final key = '${m.kind.name}|${m.leg?.name ?? '-'}';
      seen.putIfAbsent(key, () => m);
    }
    return seen.values.toList();
  }
}
